### PR TITLE
Chore/client tests codestyle docs

### PR DESCRIPTION
**Pull request by krusche**
This PR is a collaborative effort to improve client tests, code style and documentation

### Checklist
- [ ] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
We have outdated tests, JSDoc comments in the client files are missing in several places and the code style of the tests is not always in a good shape. 

This PR is an ongoing effort to collaboratively improve the client files and fix the mentioned issues:
- [x] Client code style
- [x] Fix all tests (or remove outdated ones)

It would be great if everyone can help to fix all issues in the next 2 weeks!

### CODE REVIEW

**Comment by filip-gregurevic** [comment]
The `expect` statement used by chai seems to be the reason for a lot of ` error  unused expression, expected an assignment or function call (tslint:no-unused-expression)  @typescript-eslint/tslint/config` errors. 
Since it is correct, there should be a workaround for it. 
One possibility is to install this plugin https://www.npmjs.com/package/eslint-plugin-chai-friendly
Or the rule could be disabled for all test files like: https://stackoverflow.com/questions/37558795/nice-way-to-get-rid-of-no-unused-expressions-linter-error-with-chai
In my opinion, the first way is the better one since the second option would also ignore real rule violations.
@krusche

File: src/test/javascript/spec/app/admin/user-management/user-management.component.spec.ts [hunk 1]
```diff
@@ -28,32 +32,6 @@ describe('Component Tests', () => {
...
-                    comp.ngOnInit();
-                    tick(); // simulate async
-
-                    // THEN
-                    expect(service.query).toHaveBeenCalled();
-                    expect(comp.users[0]).toEqual(jasmine.objectContaining({ id: 123 }));
-                }),
-            ));
-        });
-
```
Comment (sascha11110) [inline src/test/javascript/spec/app/admin/user-management/user-management.component.spec.ts:56]: I removed this test case as the `loadAll()` in `user-management.component.ts` is not any longer directly dependent on the `query()` method in `user.service.ts`. In this commit ngx-store was introduced: https://github.com/ls1intum/Artemis/commit/cc24454994b464fb0033bb1abb3110e577a825a3#diff-4af81ae5b1dc6f8eb6c64db0483e7951R137

**Comment by madwau** [comment]
> The `expect` statement used by chai seems to be the reason for a lot (almost all) of ` error unused expression, expected an assignment or function call (tslint:no-unused-expression) @typescript-eslint/tslint/config` errors.
> Since it is correct, there should be a workaround for it.
> One possibility is to install this plugin https://www.npmjs.com/package/eslint-plugin-chai-friendly
> Or the rule could be disabled for all test files like: https://stackoverflow.com/questions/37558795/nice-way-to-get-rid-of-no-unused-expressions-linter-error-with-chai
> The third option is to write `// eslint-disable-line no-unused-expressions` after every use of `expect`.
> In my opinion, the first way is the better one since the second option would also ignore real rule violations. (Or use the third one and copy-paste it everywhere)
> @krusche

@filip-gregurevic I agree that the first solution is the cleanest. I just added the plugin you mentioned and it seems to be work exactly as expected, i.e. chai expect statements no longer produce the `no-unused-expression` violation while the rule is still in place for non-chai statements.

I saw that you added another plugin `tslint-no-unused-expression-chai`. Did that work for you? It seemed to have no effect on the violations when I checked out the branch.

**Comment by filip-gregurevic** [comment]
> @filip-gregurevic I agree that the first solution is the cleanest. I just added the plugin you mentioned and it seems to be working exactly as expected, i.e. chai expect statements no longer produce the `no-unused-expression` violation while the rule is still in place for non-chai statements.
> 
> I saw that you added another plugin `tslint-no-unused-expression-chai`. Did that work for you? It seemed to have no effect on the violations when I checked out the branch.
> 
> For those using an IntelliJ IDE, [this tip](https://github.com/chaijs/chai/issues/41#issuecomment-455479235) also comes in handy.

For me it only worked with `tslint-no-unused-expression-chai`. There are no errors thrown in the build test either since I added it in a commit.

**Comment by madwau** [comment]
> > @filip-gregurevic I agree that the first solution is the cleanest. I just added the plugin you mentioned and it seems to be working exactly as expected, i.e. chai expect statements no longer produce the `no-unused-expression` violation while the rule is still in place for non-chai statements.
> > I saw that you added another plugin `tslint-no-unused-expression-chai`. Did that work for you? It seemed to have no effect on the violations when I checked out the branch.
> > For those using an IntelliJ IDE, [this tip](https://github.com/chaijs/chai/issues/41#issuecomment-455479235) also comes in handy.
> 
> For me it only worked with `tslint-no-unused-expression-chai`. There are no errors thrown in the build test either since I added it in a commit.

You are right, the errors were fixed in the build check when you added it.

It's probably a question whether eslint or tslint is used locally. I think for me eslint was still picking up those errors since eslint doesn't know about `tslint.json`, it only knows `.eslintrc.json`.

I'm not exactly sure why we use both tslint and eslint, maybe tslint is not a superset of eslint and eslint is complementary?

**Comment by filip-gregurevic** [comment]
> You are right, the errors were fixed in the build check when you added it.
> 
> It's probably a question whether eslint or tslint is used locally. I think for me eslint was still picking up those errors since eslint doesn't know about `tslint.json`, it only knows `.eslintrc.json`.
> 
> I'm not exactly sure why we use both tslint and eslint, maybe tslint is not a superset of eslint and eslint is complementary?

@madwau the `tslint.json` is passed to eslint in the `.eslintrc.json`. I guess this could be refactored to only use eslint or tslint. It seems the (a little bit messy) setup is due to the migration from js to ts a while back.

**Comment by madwau** [comment]
> > You are right, the errors were fixed in the build check when you added it.
> > It's probably a question whether eslint or tslint is used locally. I think for me eslint was still picking up those errors since eslint doesn't know about `tslint.json`, it only knows `.eslintrc.json`.
> > I'm not exactly sure why we use both tslint and eslint, maybe tslint is not a superset of eslint and eslint is complementary?
> 
> @madwau the `tslint.json` is passed to eslint in the `.eslintrc.json`. I guess this could be refactored to only use eslint or tslint. It seems the (a little bit messy) setup is due to the migration from js to ts a while back.

I think you are partially right. Looks like the project migrated from JS to TS on 2018-06-28. Initially, eslint was completely removed but then recently on 2020-03-23 Stephan migrated back from tslint to eslint.

**Comment by Adem-Khachnaoui** [comment]
How should we deal with warnings in mock methods? 
Is there a way to suppress all the `no-unused-vars` warnings with some kind of rule for Mock classes? 
This warning occurs at least 100 times in mock methods

**Comment by sascha11110** [comment]
> How should we deal with warnings in mock methods?
> Is there a way to suppress all the `no-unused-vars` warnings with some kind of rule for Mock classes?
> This warning occurs at least 100 times in mock methods

I think there are two possible solutions:

1. Adding `/* eslint-disable no-unused-vars */` at the top of every mock file
2. Adjust the eslint configuration for mock files. Something like that: 
```
"overrides": [{
  "files": ["mock-*.ts"],
  "rules": {
    "no-unused-vars": off
  }
}]
```

**Comment by krusche** [comment]
> > How should we deal with warnings in mock methods?
> > Is there a way to suppress all the `no-unused-vars` warnings with some kind of rule for Mock classes?
> > This warning occurs at least 100 times in mock methods
> 
> I think there are two possible solutions:
> 
> 1. Adding `/* eslint-disable no-unused-vars */` at the top of every mock file
> 2. Adjust the eslint configuration for mock files. Something like that:
> 
> ```
> "overrides": [{
>   "files": ["mock-*.ts"],
>   "rules": {
>     "no-unused-vars": off
>   }
> }]
> ```

Let’s go with the second version. We just need to make sure that all files are covered

File: package.json [hunk 2]
```diff
@@ -40,6 +40,7 @@
         "compare-versions": "3.6.0",
         "core-js": "3.6.5",
         "dompurify": "2.0.10",
+        "eslint-plugin-chai-friendly": "^0.5.0",
```
Comment (krusche) [inline package.json:43]:
```suggestion
        "eslint-plugin-chai-friendly": "0.5.0",
```

File: package.json [hunk 3]
```diff
@@ -77,6 +78,7 @@
         "sugar": "2.0.6",
         "ts-jest": "25.4.0",
         "tslib": "1.11.1",
+        "tslint-no-unused-expression-chai": "^0.1.4",
```
Comment (krusche) [inline package.json:81]:
```suggestion
        "tslint-no-unused-expression-chai": "0.1.4",
```

File: package.json [hunk 4]
```diff
@@ -176,9 +178,9 @@
         "src/{main,test}/java/**/*.java": "bash ./linting.sh"
     },
     "scripts": {
-        "prettier:format": "prettier --write 'src/{main/webapp,test}/**/*.{json,ts,css,scss,html}'",
-        "prettier:check": "prettier --check 'src/{main/webapp,test}/**/*.{json,ts,css,scss,html}'",
-        "lint": "eslint . --ext .ts",
+        "prettier:format": "prettier --write 'src/{main/webapp,test}/**/*.{json,ts,js,css,scss,html}'",
+        "prettier:check": "prettier --check 'src/{main/webapp,test}/**/*.{json,ts,js,css,scss,html}'",
+        "lint": "eslint . --ext .ts && yarn run lint:doc",
```
Comment (krusche) [inline package.json:183]:
```suggestion
        "lint": "eslint . --ext .ts",
```

**Comment by sascha11110** [comment]
List of all touched production files for testing on the test server:

```
src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts
src/main/webapp/app/account/password-reset/init/password-reset-init.component.ts
src/main/webapp/app/admin/notification-management/notification-management-update.component.ts
src/main/webapp/app/admin/notification-management/notification-management.route.ts
src/main/webapp/app/admin/user-management/user-management-update.component.ts
src/main/webapp/app/admin/user-management/user-management.route.ts
src/main/webapp/app/app.main.ts
src/main/webapp/app/assessment/assessment-warning/assessment-warning.component.ts
src/main/webapp/app/complaints/complaints-for-tutor/complaints-for-tutor.component.ts
src/main/webapp/app/complaints/list-of-complaints/list-of-complaints.component.ts
src/main/webapp/app/core/auth/user-route-access-service.ts
src/main/webapp/app/core/config/uib-pagination.config.ts
src/main/webapp/app/core/interceptor/auth-expired.interceptor.ts
src/main/webapp/app/core/interceptor/errorhandler.interceptor.ts
src/main/webapp/app/core/login/login.service.ts
src/main/webapp/app/core/websocket/websocket.service.ts
src/main/webapp/app/course/course-scores/course-scores.component.ts
src/main/webapp/app/course/manage/course-exercise-submission-result-simulation.service.ts
src/main/webapp/app/course/manage/course-group.component.ts
src/main/webapp/app/course/manage/course-management.route.ts
src/main/webapp/app/course/manage/course-management.service.ts
src/main/webapp/app/course/manage/course-update.component.ts
src/main/webapp/app/entities/lti-outcome-url.model.ts
src/main/webapp/app/entities/lti-user-id.model.ts
src/main/webapp/app/entities/single-user-notification.model.ts
src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment-dashboard.component.ts
src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.html
src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts
src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-management.route.ts
src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts
src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.service.ts
src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-conflict.component.ts
src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts
src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts
src/main/webapp/app/exercises/modeling/assess/modeling-assessment.service.ts
src/main/webapp/app/exercises/modeling/manage/example-modeling/example-modeling-submission.component.ts
src/main/webapp/app/exercises/modeling/manage/modeling-exercise-update.component.ts
src/main/webapp/app/exercises/modeling/manage/modeling-exercise.route.ts
src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts
src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts
src/main/webapp/app/exercises/programming/assess/repo-export/programming-assessment-repo-export-dialog.component.ts
src/main/webapp/app/exercises/programming/manage/programming-exercise-management-routing.module.ts
src/main/webapp/app/exercises/programming/manage/programming-exercise.component.ts
src/main/webapp/app/exercises/programming/manage/test-cases/programming-exercise-manage-test-cases.component.ts
src/main/webapp/app/exercises/programming/manage/update/programming-exercise-test-schedule-date-picker.component.ts
src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
src/main/webapp/app/exercises/programming/participate/code-editor-student-container.component.ts
src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts
src/main/webapp/app/exercises/programming/shared/code-editor/layout/code-editor-grid.component.ts
src/main/webapp/app/exercises/programming/shared/instructions-render/extensions/programming-exercise-plant-uml.extension.ts
src/main/webapp/app/exercises/programming/shared/instructions-render/extensions/programming-exercise-task.extension.ts
src/main/webapp/app/exercises/programming/shared/instructions-render/service/programming-exercise-instruction.service.ts
src/main/webapp/app/exercises/quiz/manage/apollon-diagrams/apollon-diagram-create-form.component.ts
src/main/webapp/app/exercises/quiz/manage/apollon-diagrams/apollon-diagram-list.component.ts
src/main/webapp/app/exercises/quiz/manage/apollon-diagrams/exercise-generation/quiz-exercise-generator.ts
src/main/webapp/app/exercises/quiz/manage/quiz-exercise-detail.component.ts
src/main/webapp/app/exercises/quiz/manage/quiz-exercise-popup.service.ts
src/main/webapp/app/exercises/quiz/manage/quiz-exercise.service.ts
src/main/webapp/app/exercises/quiz/manage/re-evaluate/quiz-re-evaluate-warning.component.ts
src/main/webapp/app/exercises/quiz/manage/statistics/quiz-point-statistic/quiz-point-statistic.component.ts
src/main/webapp/app/exercises/quiz/participate/quiz-participation.component.ts
src/main/webapp/app/exercises/quiz/shared/questions/drag-and-drop-question/drag-and-drop-question.component.ts
src/main/webapp/app/exercises/shared/dashboards/tutor/tutor-exercise-dashboard.component.ts
src/main/webapp/app/exercises/shared/exercise-headers/difficulty-badge.component.ts
src/main/webapp/app/exercises/shared/exercise-headers/header-exercise-page-with-details.component.ts
src/main/webapp/app/exercises/shared/exercise-headers/header-participation-page.component.ts
src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint-update.component.ts
src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint.component.ts
src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint.route.ts
src/main/webapp/app/exercises/shared/exercise/exercise-popup.service.ts
src/main/webapp/app/exercises/shared/participation-submission/participation-submission-popup.service.ts
src/main/webapp/app/exercises/shared/participation/participation.service.ts
src/main/webapp/app/exercises/shared/result/result.service.ts
src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts
src/main/webapp/app/exercises/shared/team/team-exercise-search/team-exercise-search.component.ts
src/main/webapp/app/exercises/shared/team/team-owner-search/team-owner-search.component.ts
src/main/webapp/app/exercises/shared/team/team-student-search/team-student-search.component.ts
src/main/webapp/app/exercises/shared/team/team-update-dialog/team-delete-button.component.ts
src/main/webapp/app/exercises/shared/team/team.component.html
src/main/webapp/app/exercises/shared/team/team.component.ts
src/main/webapp/app/exercises/shared/team/teams.component.ts
src/main/webapp/app/exercises/text/assess-new/text-assessment-area/text-assessment-area.component.ts
src/main/webapp/app/exercises/text/assess-new/text-submission-assessment.component.ts
src/main/webapp/app/exercises/text/assess-new/text-submission-assessment.route.ts
src/main/webapp/app/exercises/text/assess-new/textblock-feedback-editor/textblock-feedback-editor.component.ts
src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.ts
src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-update.component.ts
src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise.route.ts
src/main/webapp/app/exercises/text/participate/text-editor.component.ts
src/main/webapp/app/exercises/text/participate/text-submission.service.ts
src/main/webapp/app/guided-tour/guided-tour.component.ts
src/main/webapp/app/guided-tour/guided-tour.service.ts
src/main/webapp/app/guided-tour/tours/course-overview-tour.ts
src/main/webapp/app/home/home.component.ts
src/main/webapp/app/lecture/lecture-update.component.ts
src/main/webapp/app/lecture/lecture.route.ts
src/main/webapp/app/overview/course-lectures/course-lecture-row.component.html
src/main/webapp/app/overview/course-lectures/course-lecture-row.component.ts
src/main/webapp/app/overview/course-lectures/course-lectures.component.ts
src/main/webapp/app/overview/course-statistics/course-statistics.component.ts
src/main/webapp/app/overview/student-questions/student-question-row.component.ts
src/main/webapp/app/shared/category-selector/category-selector.component.ts
src/main/webapp/app/shared/components/confirm-autofocus-button.component.ts
src/main/webapp/app/shared/data-table/data-table.component.ts
src/main/webapp/app/shared/date-time-picker/date-time-picker.component.ts
src/main/webapp/app/shared/layouts/profiles/page-ribbon.component.ts
src/main/webapp/app/shared/markdown-editor/commands/fullscreen.command.ts
src/main/webapp/app/shared/markdown-editor/commands/katex.command.ts
src/main/webapp/app/shared/pipes/format-date.pipe.ts
src/main/webapp/app/shared/result/result-detail.component.ts
src/main/webapp/app/shared/util/security-util.ts
```

**Comment by kloessst** [comment]
Went through the production files. Looks fine except some components like
code-editor-grid.component.ts, team-delete-button.component.ts, category-selector.component.ts use eslint-disable-next-line @typescript-eslint/no-unused-vars instead of deleting the unused variables. Any good reason for that?

**Comment by madwau** [comment]
Looked through the production files and didn't spot any issues. Testing also went fine.

**Pull request merged**